### PR TITLE
Add new `UNSTABLE_INTRINSICS_WITH_STABLE_WRAPPER` lint

### LINKED ISSUE #2997: Lint against calling intrinsics that have stable or safe wrappers
There are some intrinsics that there should be no reason to call. For example [`core::intrinsics::size_of`](https://doc.rust-lang.org/core/intrinsics/fn.size_of.html). This intrinsic is exposed by the safe function [`core::mem::size_of`](https://doc.rust-lang.org/core/mem/fn.size_of.html) which should always be called instead.

I would like to suggest a Clippy lint that notices calls to no-reason-to-call intrinsics and recommends the right stable or safe wrapper to call instead.

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #2997.

To prevent having to keep the list of intrinsics with stable counterpart up-to-date on clippy side, I instead check if the documentation contains the sentence `"does not have a stable counterpart"`. To prevent running this check more than once, I store all the non-matching intrinsics `DefId` into a `static`.

r? @llogiq 

changelog: Add new `UNSTABLE_INTRINSICS_WITH_STABLE_WRAPPER` lint

### CODE REVIEW

**Comment by GuillaumeGomez** [comment]
Interestingly enough, we cannot differentiate between the intrinsics item and not the reexported one (like for `std::ptr::write_bytes`). I'll try to extract more information to ensure which call is actually performed.

**Comment by GuillaumeGomez** [comment]
Took me a while to realize that the error was coming from the doc example... ^^'

PR is now ready!

File: tests/ui/unstable_intrinsics_with_stable_wrapper.stderr [hunk 1]
```diff
@@ -0,0 +1,17 @@
+error: consider using the stable counterpart mentioned in the documentation (https://rust-lang.org/stable/core/intrinsics/fn.add_with_overflow.html)
```
Comment (Alexendoo) [inline tests/ui/unstable_intrinsics_with_stable_wrapper.stderr:1]: The URLs 404

Comment (GuillaumeGomez) [reply tests/ui/unstable_intrinsics_with_stable_wrapper.stderr:1]: Arf. Forgot the `doc.`... Thanks!

File: clippy_lints/src/unstable_intrinsics_with_stable_wrapper.rs [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+}
+
+fn emit_if_is_unstable_intrinsic(
+    cx: &LateContext<'_>,
+    def_id: DefId,
+    expr_span: Span,
+    fn_name: &str,
+    segments: &[PathSegment<'_>],
+) {
+    static FNS: OnceLock<Vec<DefId>> = OnceLock::new();
```
Comment (Centri3) [inline clippy_lints/src/unstable_intrinsics_with_stable_wrapper.rs:59]: Why not a `DefIdMap`?

File: clippy_lints/src/unstable_intrinsics_with_stable_wrapper.rs [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+    let fns = FNS.get_or_init(|| {
+        let mod_def_id = def_path_def_ids(cx, &["core", "intrinsics"]).next().unwrap();
+        cx.tcx
+            .module_children(mod_def_id)
+            .iter()
+            .filter_map(|child| match child.res {
+                Res::Def(DefKind::Fn, fn_def_id) => Some(fn_def_id),
+                _ => None,
+            })
+            .filter(|fn_def_id| !get_doc(cx.tcx, *fn_def_id).contains("does not have a stable counterpart"))
```
Comment (Centri3) [inline clippy_lints/src/unstable_intrinsics_with_stable_wrapper.rs:69]: Some intrinsics don't have this. For example `is_val_statically_known` or `const_eval_select`

**Review by Centri3 — COMMENTED** [review]
Tbh I think this should be a rustc lint instead, that way there can be a "stabilized counterpart" attribute it checks for instead.

File: clippy_lints/src/unstable_intrinsics_with_stable_wrapper.rs (hunk 3 above)
Comment (Centri3) [reply clippy_lints/src/unstable_intrinsics_with_stable_wrapper.rs:69]: It's probably best to at least check for "the stabilized version of this intrinsic is" after this too

File: tests/ui/transmute.stderr [hunk 4]
```diff
@@ -1,5 +1,5 @@
 error: transmute from a reference to a pointer
```
Comment (llogiq) [inline tests/ui/transmute.stderr:1]: I think this lint should be allowed in other tests.

File: clippy_lints/src/unstable_intrinsics_with_stable_wrapper.rs (hunk 2 above)
Comment (llogiq) [reply clippy_lints/src/unstable_intrinsics_with_stable_wrapper.rs:59]: I don't think having a `Vec` will be a problem here, we don't have that many intrinsics anyway.

Comment (Centri3) [reply clippy_lints/src/unstable_intrinsics_with_stable_wrapper.rs:59]: I guess. A benchmark is always handy, I just think it's cleaner at least and more greppable

**Comment by J-ZhengLi** [comment]
Hey @GuillaumeGomez , this is a ping from triage. There hasn't been any activity for some time, perhaps it was forgotten?

@rustbot author

**Comment by GuillaumeGomez** [comment]
I'll close it because it should likely be a rustc lint directly.

**Pull request closed without merging**